pkg/consuldp/metrics.go

```diff
@@ -0,0 +1,96 @@
+package consuldp
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+)
+
+const (
+	envoyMetricsUrl        = "http://127.0.0.1:19000/stats/prometheus"
+	metricsBackendBindPort = "20100"
+	metricsBackendBindAddr = "127.0.0.1:" + metricsBackendBindPort
+)
+
+func (cdp *ConsulDataplane) setupMetricsServer() {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/stats/prometheus", cdp.mergedMetricsHandler)
+	cdp.metricsServer = &metricsServer{
+		httpServer: &http.Server{
+			Addr:    metricsBackendBindAddr,
+			Handler: mux,
+		},
+		client: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+		exitedCh: make(chan struct{}),
+	}
+}
+
+func (cdp *ConsulDataplane) startMetricsServer() {
+	cdp.logger.Info("starting metrics server", "address", cdp.metricsServer.httpServer.Addr)
+	defer close(cdp.metricsServer.exitedCh)
+	err := cdp.metricsServer.httpServer.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
+		cdp.logger.Error("failed to serve metrics requests", "error", err)
+	}
+}
+
+func (cdp *ConsulDataplane) stopMetricsServer() {
+	if cdp.metricsServer != nil && cdp.metricsServer.httpServer != nil {
+		cdp.logger.Debug("stopping metrics server")
+		err := cdp.metricsServer.httpServer.Close()
+		if err != nil {
+			cdp.logger.Warn("error while closing metrics server", "error", err)
+		}
+	}
+}
+
+func (cdp *ConsulDataplane) metricsServerExited() <-chan struct{} {
+	return cdp.metricsServer.exitedCh
+}
+
+// mergedMetricsHandler responds with merged metrics from multiple sources:
+// Consul Dataplane, Envoy and (optionally) the service/application. The Envoy
+// and service metrics are scraped synchronously during the handling of this
+// request.
+func (cdp *ConsulDataplane) mergedMetricsHandler(rw http.ResponseWriter, _ *http.Request) {
+	cdp.logger.Debug("scraping Envoy metrics", "url", envoyMetricsUrl)
+	if err := cdp.scrapeMetrics(rw, envoyMetricsUrl); err != nil {
+		cdp.scrapeError(rw, envoyMetricsUrl, err)
+		return
+	}
+	telem := cdp.cfg.Telemetry
+	if telem == nil || telem.Prometheus.ServiceMetricsURL == "" {
+		return
+	}
+	url := telem.Prometheus.ServiceMetricsURL
+	cdp.logger.Debug("scraping service metrics", "url", url)
+	if err := cdp.scrapeMetrics(rw, url); err != nil {
+		cdp.scrapeError(rw, url, err)
+		return
+	}
+}
+
+// scrapeMetrics fetches metrics from the given url and copies them to the response.
+func (cdp *ConsulDataplane) scrapeMetrics(rw http.ResponseWriter, url string) error {
+	resp, err := cdp.metricsServer.client.Get(url)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
```

Comment:
nit: if we're just matching on 200, then we could use:
```suggestion
	if resp.StatusCode != http.StatusOK {
```

That said, I think consul-k8s handles [any 2XX code](https://github.com/hashicorp/consul-k8s/blob/168ccba9eb3d31b3ccd7fdda14f51d31dc853172/control-plane/subcommand/consul-sidecar/command.go#L294), which we should probably do here too.

Reply:
I also updated this to make sure the response body is closed.

Another couple of differences with k8s as I'm looking again:

- In case of an http error (non-2xx status code), this does not forward along the response body from the metrics source (Envoy or service), [whereas k8s does](https://github.com/hashicorp/consul-k8s/blob/168ccba9eb3d31b3ccd7fdda14f51d31dc853172/control-plane/subcommand/consul-sidecar/command.go#L296). 
- k8s writes an additional metric in the response: `consul_merged_service_metrics_success`

Is it important to closely match the k8s behavior here? cc @ishustava for thoughts on this